cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -0,0 +1,127 @@
+		// The old manifest will be moved here; into a subfolder of the temporary directory
+		// If a rollback is needed, these manifests will be put back to where they where initially
+		backupManifestPath := constants.GetStaticPodFilepath(component, backupManifestPath)
+		// The new, upgraded manifest will be written here
+		newManifestPath := constants.GetStaticPodFilepath(component, tmpdir)
+
+		// Store the backup path in the recover list. If something goes wrong now, this component will be rolled back.
+		recoverManifests[component] = backupManifestPath
+
+		// Move the old manifest into the old-manifests directory
+		if err := os.Rename(currentManifestPath, backupManifestPath); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		// Move the new manifest into the manifests directory
+		if err := os.Rename(newManifestPath, currentManifestPath); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		fmt.Printf("[upgrade/staticpods] Moved upgraded manifest to %q and backuped old manifest to %q\n", currentManifestPath, backupManifestPath)
+		fmt.Println("[upgrade/staticpods] Waiting for the kubelet to restart the component")
+
+		// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+		// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+		// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+		// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+		if err := kubeadmutil.WaitForStaticPodControlPlaneHashChange(client, waitForComponentReadyTimeout, cfg.NodeName, component, beforePodHashMap[component]); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		// Wait for the static pod component to come up and register itself as a mirror pod
+		if err := kubeadmutil.WaitForPodsWithLabel(client, waitForComponentReadyTimeout, os.Stdout, "component="+component); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	}
+	return nil
+}
+
+func rollbackOldManifests(oldManifests map[string]string, origErr error) error {
+	errs := []error{origErr}
+	for component, backupPath := range oldManifests {
+		// Where we should put back the backuped manifest
+		realManifestPath := constants.GetStaticPodFilepath(component, constants.GetStaticPodDirectory())
+
+		// Move the backup manifest back into the manifests directory
+		err := os.Rename(backupPath, realManifestPath)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
```

Comment:
is it worth waiting until the recovered pod is healthy? because if we put it back and it doesn't come back online, the user has _another_ thing to worry about. A failed upgrade is one thing, but a failed upgrade which brings down your cluster is another

Reply:
I don't think there is any upgrade procedure that is totally 100% safe to take, as the code is written by humans ;)
Anyway, rolling back is on a best-effort; do you mean there would be anything more we could do?
What would we do if the old manifest _doesn't_ come up?

Reply:
I definitely haven't been involved enough in the design or these discussions, but have a long personal history of doing large migrations that were carefully designed to be safe. For instance, if you're doing a data migration from store A to store B, a safe plan could be something like the following:

1) Keep all read operations using A, but change writes so they propagate to both A and B simultaneously.
2) Backfill any data that exists in A and is missing in B into B.
3) Run a consistency checker that is constantly probing to make sure there aren't discrepancies between data in A and B.
4) Change read operations to read from A and B simultaneously. Always return the value from A, but create a silent report if the values of A and B didn't match.
5) When enough time has passed and you're certain that B is healthy and up to date, change reads to only read from B.
6) Change writes to only write to B.
7) Completely retire A.

It might seem overly complex, especially since each step is a different launch that is incrementally getting you closer to the full migration, but it has the nice property that at any point, if something goes wrong, you can always stop the migration or roll back to previous steps without any data loss. There is no single "flip over to B" step that has a window of uncertainty or potential data loss if things don't go well. You never move onto the next step until the previous step has proved successful, and you can stop safely at any time, and move backwards in the plan at any time. Contrast this to the fast-but-easy plan of taking scheduled downtime to flip everything over to point to B all at once, and crossing your fingers that it all works when you bring the system back online.

It would be great if our upgrade plan had similar properties, so that kubeadm or any random component could crash at any time during the upgrade, and we would never leave the cluster in an unusable state where human intervention would be required to fix it. But, I don't know how feasible that is with the mechanisms we have today, and don't have specific suggestions on how to get there now, so this is just a general grumbling that I hope we can figure out how to do better.

If I have time to offer concrete suggestions for how to do these steps differently, I'll let you know. :-)